File: apps/web/static/js/blocks/ribbonmodel.ts
import { t, type MessageKey } from '../i18n.js';
import { hhmmToMin, isPublicSegment, minToHhmm } from './cursor.js';
// ribbonmodel.js — PURE mapping of a `/swim` option/status → a ribbon render state.
//
// No canvas, no DOM (plan Risk #2: keep the drawable logic testable without a
// browser). The board's canvas renderer consumes these plain objects; every visual
// decision (dashed vs dotted, thickness, pinch, sheath, colour family) is decided
// HERE and unit-tested against saved `/swim` fixtures.
//
// The three terminal states are never merged (product invariant): a `closed` status
// is a DASHED ribbon carrying its reason; a schedule-less status (`awaiting_scrape`
// or `no_source`) is a DOTTED ghost; an option with no published lane split is its own
// "not published" ribbon — none of them collapses into another.

// access class name (`type(session.access).__name__`) → colour-family key. The board
// maps each key to a token (see blocks.css `.fam-*`); the key itself carries no hex.
export const ACCESS_FAMILY = Object.freeze({
  PublicSwim: 'public',
  LaneSwim: 'lane',
  FamilyTime: 'family',
  WomenOnly: 'women',
  SeniorsOnly: 'seniors',
  AdultsOnly: 'adults',
  SchoolReserved: 'school',
  ClubReserved: 'club',
  // The school-pool vocabulary (school-access-vocabulary S1). Each gets its OWN key rather
  // than falling to 'other': 'other' paints with the public-swim colour (ribbonrender's
  // `pal.other = read('fam-public')`), which is exactly the "looks open to you" lie the
  // eligibility fallback used to tell.
  GirlsOnly: 'girls',
  GenderDiverse: 'diverse',
  AccompaniedChildren: 'accompanied',
});

/** A `/swim` option as the ribbon model reads it. */
export interface RibbonOption {
  access?: string;
  start?: string;
  end?: string;
  facility?: string;
  basin?: string;
  lane_timeline?: { segments?: RibbonTimelineSegment[] } | null;
  /** `OptionOut.lane_day_view` (S2): WHICH lane and WHOSE, for the lane stack. */
  lane_day_view?: RibbonDayView | null;
  /** `OptionOut.lane_best_public` (S2): the best public window, ALREADY bounded by this
   *  option's own session hours server-side — the stack must not re-widen it. */
  lane_best_public?: RibbonPublicWindow | null;
  [k: string]: unknown;
}

/** One owner's hold on one lane, as `/swim` serves it (`LaneSegmentOut`). */
export interface RibbonLaneSegment {
  start: string;
  end: string;
  access?: string;
  owner?: string | null;
  [k: string]: unknown;
}

/** One lane's whole weekday (`LaneStripOut`). */
export interface RibbonLaneStrip {
  lane: number;
  segments?: RibbonLaneSegment[];
  [k: string]: unknown;
}

/** `LaneDayViewOut` — the basin's per-lane WEEKDAY (not the session window). */
export interface RibbonDayView {
  weekday?: number;
  lane_count: number;
  strips?: RibbonLaneStrip[];
  [k: string]: unknown;
}

/** `PublicWindowOut` — the session's "best time to come". */
export interface RibbonPublicWindow {
  start: string;
  end: string;
  public_lanes: number;
}

/** One block of one lane of the STACK: a drawable, session-clipped hold. */
export interface RibbonStackBlock {
  start: string;
  end: string;
  public: boolean;
  owner: string | null;
}

/** One lane sub-row of the stack. Always present for lanes `1..lane_count`, even when the
 *  lane holds nothing inside the session — an empty sub-row is a lane nobody has taken,
 *  which is a different fact from a lane we know nothing about. */
export interface RibbonStackLane {
  lane: number;
  segments: RibbonStackBlock[];
}

export interface RibbonTimelineSegment {
  start: string;
  end: string;
  lane_count: number;
  public_lanes: number;
  [k: string]: unknown;
}

export interface RibbonStatus {
  facility?: string;
  status?: string;
  detail?: string | null;
  closure_code?: string | null;
  detail_params?: Record<string, string>;
}

/** One drawable band on a board row. */
export interface Ribbon {
  kind: string;
  variant?: string;
  family?: string;
  style?: string;
  [k: string]: unknown;
}

/** The i18n key for the "lane split not published" label — the one ribbon field whose
 *  rendered form is locale-dependent. Exported so the golden fixture can carry the key
 *  rather than a translation. */
export const NO_SPLIT_LABEL_KEY: MessageKey = 'insight.noSplit.label';

/** access class name → colour-family key ('other' for an unknown type). */
export function accessFamily(access: string): string {
  return (ACCESS_FAMILY as Record<string, string>)[access] || 'other';
}

// Public fraction of a lane-timeline segment: how much of the ribbon is open to you.
// Explicit 0 when the basin has no lanes recorded (avoid NaN → the ribbon pinches shut).
function publicFraction(seg: RibbonTimelineSegment): number {
  return seg.lane_count > 0 ? seg.public_lanes / seg.lane_count : 0;
}

/**
 * laneStackFor(dayView, start, end) → one sub-row per lane `1..lane_count`, each carrying
 * only the holds that fall inside THIS option's hours.
 *
 * The clip is not cosmetic. `lane_day_view` spans the whole WEEKDAY, while a ribbon is one
 * SESSION: Oerlikon's 06:00–08:00 and 08:00–21:30 options share a basin and therefore
 * share a day view, so an unclipped stack would paint each option's ribbon across the
 * whole day — two ribbons claiming hours neither session covers, drawn over each other.
 * Clipping mirrors what S2 already did server-side for `lane_best_public`.
 *
 * A lane with nothing inside the window keeps its (empty) sub-row: "nobody holds this
 * lane" is a fact, and dropping the row would silently renumber the lanes below it.
 */
export function laneStackFor(
  dayView: RibbonDayView,
  start: string | undefined,
  end: string | undefined,
): RibbonStackLane[] {
  const lo = hhmmToMin(start ?? '');
  const hi = hhmmToMin(end ?? '');
  const byLane = new Map<number, RibbonLaneSegment[]>();
  for (const strip of dayView.strips ?? []) {
    byLane.set(Number(strip.lane), strip.segments ?? []);
  }
  const lanes: RibbonStackLane[] = [];
  for (let lane = 1; lane <= dayView.lane_count; lane += 1) {
    const segments: RibbonStackBlock[] = [];
    for (const seg of byLane.get(lane) ?? []) {
      const s = Math.max(lo, hhmmToMin(seg.start));
      const e = Math.min(hi, hhmmToMin(seg.end));
      if (!(e > s)) continue; // wholly outside this session (or zero-length)
      segments.push({
        start: minToHhmm(s),
        end: minToHhmm(e),
        public: isPublicSegment(seg),
        owner: typeof seg.owner === 'string' && seg.owner ? seg.owner : null,
      });
    }
    lanes.push({ lane, segments });
  }
  return lanes;
}

/**
 * optionRibbon(option) → a ribbon render state for a `/swim` option.
 *   - with `lane_day_view`  → a LANE STACK: one sub-row per lane, public vs reserved,
 *     the owner carried on each reserved block, and the session's best-public window
 *     (`option.lane_best_public`, absent when the server sent null).
 *   - with `lane_timeline` only → a FILLED ribbon; each segment's `thickness` is the
 *     public fraction (public_lanes/lane_count) and it is `pinched` wherever any
 *     lane is reserved (reserved_lanes>0); `sheath:true` draws the capacity envelope.
 *   - with neither → a "lane split not published" ribbon (`sheath:false`).
 * The colour `family` is set in every case.
 *
 * The three are NEVER merged (invariant I5): the published universe is closed at 8 lane
 * sheets, so most pools will never have a stack, and a pool with no plan must not read as
 * a pool with no free lanes.
 */
export function optionRibbon(option: RibbonOption): Ribbon {
  const family = accessFamily(option.access ?? '');
  const base = {
    kind: 'option',
    family,
    access: option.access,
    facility: option.facility,
    basin: option.basin,
    start: option.start,
    end: option.end,
  };
  const dayView = option.lane_day_view;
  // Hours are required to CLIP the day view to this session, so an option missing them
  // cannot be stacked: a stack with no window would paint every lane empty, i.e. "nothing
  // free", which is the one thing a missing fact must never look like (I5).
  const hasHours =
    Number.isFinite(hhmmToMin(option.start ?? '')) && Number.isFinite(hhmmToMin(option.end ?? ''));
  if (dayView && Number(dayView.lane_count) > 0 && Array.isArray(dayView.strips) && hasHours) {
    return {
      ...base,
      variant: 'lanestack',
      style: 'solid',
      sheath: true,
      lane_count: Number(dayView.lane_count),
      strips: laneStackFor(dayView, option.start, option.end),
      // ABSENT, not zero-width, when the server has no window for this session: the band
      // is a claim ("come then"), and a null one is no claim at all.
      ...(option.lane_best_public ? { best_public: option.lane_best_public } : {}),
    };
  }
  const timeline = option.lane_timeline;
  if (timeline && Array.isArray(timeline.segments) && timeline.segments.length > 0) {
    return {
      ...base,
      variant: 'lanes',
      style: 'solid',
      sheath: true,
      segments: timeline.segments.map((seg) => ({
        start: seg.start,
        end: seg.end,
        thickness: publicFraction(seg),
        pinched: Number(seg.reserved_lanes) > 0,
        lane_count: seg.lane_count,
        public_lanes: seg.public_lanes,
        reserved_lanes: Number(seg.reserved_lanes),
        partial: seg.partial,
      })),
    };
  }
  return {
    ...base,
    variant: 'unpublished',
    style: 'solid',
    sheath: false,
    label: t(NO_SPLIT_LABEL_KEY),
    // The KEY beside the rendered label. `label` is locale-dependent output, so it is the one
    // field of a ribbon that cannot appear in a cross-client golden fixture without pinning
    // this suite's locale into it; the key is the thing both clients actually agree on, and
    // the iOS port renders it through its own catalog.
    label_key: NO_SPLIT_LABEL_KEY,
  };
}

/**
 * statusRibbon(status) → a ribbon render state for a `/swim` status entry.
 *   - status === 'closed'                       → a DASHED closed ribbon carrying `detail`.
 *   - status === 'awaiting_scrape' | 'no_source' → a DOTTED ghost ribbon (unknown ≠ closed).
 * Any other status label falls back to the ghost/unknown ribbon (never to closed).
 */
export function statusRibbon(status: RibbonStatus): Ribbon {
  // Carry the S4 closure code alongside the prose so the CANVAS ribbon and the label
  // column render the same fact the same way — they used to diverge (label translated,
  // ribbon still German), which is worse than either alone.
  const base = {
    kind: 'status',
    facility: status.facility,
    detail: status.detail,
    closure_code: status.closure_code,
    detail_params: status.detail_params,
  };
  if (status.status === 'closed') {
    return { ...base, variant: 'closed', style: 'dashed', family: 'closed' };
  }
  // A schedule-less status (awaiting_scrape / no_source) → a DOTTED ghost; carry the status value
  // so the canvas can render its SPECIFIC freshness label, never a merged bucket.
  return { ...base, status: status.status, variant: 'ghost', style: 'dotted', family: 'unknown' };
}

/**
 * ribbonsFor(row) → every ribbon for one board row, statuses first (background
 * closed/ghost states) then the option ribbons on top. `row` is
 * `{ options: OptionOut[], statuses: StatusOut[] }`.
 */
export function ribbonsFor(row: {
  options?: RibbonOption[];
  statuses?: RibbonStatus[];
}): Ribbon[] {
  const statuses = (row.statuses || []).map(statusRibbon);
  const options = (row.options || []).map(optionRibbon);
  return [...statuses, ...options];
}

File: apps/ios/Sources/SwimZHKit/RibbonModel.swift
// RibbonModel.swift — the port of `apps/web/static/js/blocks/ribbonmodel.ts`.
//
// A ribbon is the PURE mapping from one option (or one status) to a drawable state: which
// variant, which colour family, which line style, how thick, and — for a basin with a parsed
// Belegungsplan — the per-lane stack clipped to the session's own hours. No canvas, no SwiftUI:
// every visual decision is made here so it can be tested, and the renderer only obeys.
//
// THE THREE TERMINAL STATES ARE NEVER MERGED, which is this file's whole product invariant:
//   * `closed` — the source says shut. A DASHED ribbon carrying the reason.
//   * `awaiting_scrape` / `no_source` — we do not know its hours. A DOTTED ghost.
//   * `unpublished` — the pool's hours are known, its LANE SPLIT is not. Its own ribbon.
// A pool with no lane plan must never read as a pool with no free lanes, and an unknown
// schedule must never read as a closure. The published universe is eight lane sheets, so most
// pools will never have a stack, and collapsing the third state into the second would mislabel
// almost the whole city.
//
// WHY THIS IS A PORT AND NOT A SECOND IMPLEMENTATION. `blocks/fixtures/ribbon_golden.json` is
// emitted by `ribbonmodel.test.ts` and replayed by `RibbonModelTests`: the browser and the
// phone are held to the same ribbon for the same input, entry by entry. One field is
// deliberately absent from that contract — the rendered `label`, which is `t(...)` output and
// therefore locale-dependent. Its KEY (`labelKey`) is carried instead, and S4 renders it.

import Foundation

/// Access class name → colour-family key. The key carries no colour: `Theme.swift` maps it to
/// an Asset Catalog entry, exactly as the web maps it to a `.fam-*` token.
///
/// The three school-pool arms get their OWN keys rather than falling to `other`, and that is a
/// correctness rule rather than a palette preference: `other` paints in the public-swim colour,
/// so a girls-only session would be drawn in the open-to-all family — the "looks open to you"
/// lie the whole eligibility vocabulary exists to prevent.
public let accessFamilies: [String: String] = [
  "PublicSwim": "public",
  "LaneSwim": "lane",
  "FamilyTime": "family",
  "WomenOnly": "women",
  "SeniorsOnly": "seniors",
  "AdultsOnly": "adults",
  "SchoolReserved": "school",
  "ClubReserved": "club",
  "GirlsOnly": "girls",
  "GenderDiverse": "diverse",
  "AccompaniedChildren": "accompanied",
]

/// The family for an access class name; `other` for one this binary has never heard of — which
/// a store built by a newer export can carry.
public func accessFamily(_ access: String) -> String {
  accessFamilies[access] ?? "other"
}

/// The i18n key for the "lane split not published" label. The KEY travels in the golden; the
/// sentence is S4's.
public let noSplitLabelKey = "insight.noSplit.label"

/// One drawable band. Optional-as-absent throughout: a field that is nil is a field the ribbon
/// does not have, which is exactly how the browser's object literal behaves and exactly what
/// the golden records.
public struct Ribbon: Equatable, Sendable, Codable {
  public let kind: String
  public let variant: String
  public let style: String
  public let family: String
  public let access: String?
  public let facility: String?
  public let basin: String?
  public let start: String?
  public let end: String?
  public let sheath: Bool?
  public let laneCount: Int?
  public let strips: [RibbonStackLane]?
  public let segments: [RibbonSegment]?
  public let bestPublic: RibbonPublicWindow?
  public let labelKey: String?
  public let detail: String?
  public let closureCode: String?
  public let detailParams: [String: String]?
  public let status: String?

  enum CodingKeys: String, CodingKey {
    case kind
    case variant
    case style
    case family
    case access
    case facility
    case basin
    case start
    case end
    case sheath
    case laneCount = "lane_count"
    case strips
    case segments
    case bestPublic = "best_public"
    case labelKey = "label_key"
    case detail
    case closureCode = "closure_code"
    case detailParams = "detail_params"
    case status
  }

  /// The session's window, when the ribbon has one. The wire keeps `"HH:MM"` because that is
  /// what the golden compares; the renderer wants minutes.
  public var window: TimeWindow? {
    guard let start, let end, let lower = TimeOfDay(hhmm: start), let upper = TimeOfDay(hhmm: end)
    else { return nil }
    return TimeWindow(start: lower, end: upper)
  }
}

/// One segment of a lane-timeline ribbon: the split over a sub-window of the session.
public struct RibbonSegment: Equatable, Sendable, Codable {
  public let start: String
  public let end: String
  /// `public_lanes / lane_count` — the fraction of capacity open to you, and the ribbon's
  /// thickness about its mid-line.
  public let thickness: Double
  /// Any lane is reserved, so the ribbon is drawn narrower than the fraction alone: a
  /// second, non-colour channel for "someone else is in the water".
  public let pinched: Bool
  public let laneCount: Int
  public let publicLanes: Int
  public let reservedLanes: Int
  /// The count may be incomplete — from `PlanCoverage.unresolved_lanes`.
  public let partial: Bool?

  enum CodingKeys: String, CodingKey {
    case start
    case end
    case thickness
    case pinched
    case laneCount = "lane_count"
    case publicLanes = "public_lanes"
    case reservedLanes = "reserved_lanes"
    case partial
  }
}

/// One lane's sub-row of a stack, clipped to the session.
public struct RibbonStackLane: Equatable, Sendable, Codable, Identifiable {
  public let lane: Int
  public let segments: [RibbonStackBlock]

  public var id: Int { lane }
}

/// One drawable hold inside a stack sub-row.
public struct RibbonStackBlock: Equatable, Sendable, Codable, Identifiable {
  public let start: String
  public let end: String
  /// Open to the public. `public` is a Swift keyword, so the property is renamed and the
  /// coding key keeps the wire's spelling.
  public let isPublic: Bool
  public let owner: String?

  public var id: String { "\(start)|\(end)" }

  enum CodingKeys: String, CodingKey {
    case start
    case end
    case isPublic = "public"
    case owner
  }
}

/// The "best time to come" band, already bounded by the session server-side.
public struct RibbonPublicWindow: Equatable, Sendable, Codable {
  public let start: String
  public let end: String
  public let publicLanes: Int

  enum CodingKeys: String, CodingKey {
    case start
    case end
    case publicLanes = "public_lanes"
  }
}

// MARK: - The inputs

/// One option, as the ribbon model reads it. The app builds this from a `SwimOption` plus the
/// lane derivations; the golden test decodes it straight from a `/swim` payload, which is what
/// keeps the two clients comparable at all.
public struct RibbonOptionInput: Equatable, Sendable, Decodable {
  public let access: String?
  public let start: String?
  public let end: String?
  public let facility: String?
  public let basin: String?
  public let laneTimeline: RibbonTimelineInput?
  public let laneDayView: RibbonDayViewInput?
  public let laneBestPublic: RibbonPublicWindow?

  enum CodingKeys: String, CodingKey {
    case access
    case start
    case end
    case facility
    case basin
    case laneTimeline = "lane_timeline"
    case laneDayView = "lane_day_view"
    case laneBestPublic = "lane_best_public"
  }

  public init(
    access: String?,
    start: String?,
    end: String?,
    facility: String?,
    basin: String?,
    laneTimeline: RibbonTimelineInput? = nil,
    laneDayView: RibbonDayViewInput? = nil,
    laneBestPublic: RibbonPublicWindow? = nil
  ) {
    self.access = access
    self.start = start
    self.end = end
    self.facility = facility
    self.basin = basin
    self.laneTimeline = laneTimeline
    self.laneDayView = laneDayView
    self.laneBestPublic = laneBestPublic
  }
}

public struct RibbonTimelineInput: Equatable, Sendable, Decodable {
  public struct Segment: Equatable, Sendable, Decodable {
    public let start: String
    public let end: String
    public let laneCount: Int
    public let publicLanes: Int
    public let reservedLanes: Int?
    public let partial: Bool?

    enum CodingKeys: String, CodingKey {
      case start
      case end
      case laneCount = "lane_count"
      case publicLanes = "public_lanes"
      case reservedLanes = "reserved_lanes"
      case partial
    }

    public init(
      start: String, end: String, laneCount: Int, publicLanes: Int, reservedLanes: Int?,
      partial: Bool?
    ) {
      self.start = start
      self.end = end
      self.laneCount = laneCount
      self.publicLanes = publicLanes
      self.reservedLanes = reservedLanes
      self.partial = partial
    }
  }

  public let segments: [Segment]?

  public init(segments: [Segment]?) {
    self.segments = segments
  }
}

public struct RibbonDayViewInput: Equatable, Sendable, Decodable {
  public struct Segment: Equatable, Sendable, Decodable {
    public let start: String
    public let end: String
    public let access: String?
    public let owner: String?

    public init(start: String, end: String, access: String?, owner: String?) {
      self.start = start
      self.end = end
      self.access = access
      self.owner = owner
    }
  }

  public struct Strip: Equatable, Sendable, Decodable {
    public let lane: Int
    public let segments: [Segment]?

    public init(lane: Int, segments: [Segment]?) {
      self.lane = lane
      self.segments = segments
    }
  }

  public let weekday: Int?
  public let laneCount: Int
  public let strips: [Strip]?

  enum CodingKeys: String, CodingKey {
    case weekday
    case laneCount = "lane_count"
    case strips
  }

  public init(weekday: Int?, laneCount: Int, strips: [Strip]?) {
    self.weekday = weekday
    self.laneCount = laneCount
    self.strips = strips
  }
}

/// One status, as the ribbon model reads it.
public struct RibbonStatusInput: Equatable, Sendable, Decodable {
  public let facility: String?
  public let status: String?
  /// The SERVER's own prose for the state, when the input came off a `/swim` payload. It is
  /// English in one branch and curated German in the other (CLAUDE.md), so nothing user-facing
  /// reads it any more; it survives because the golden fixture carries it and dropping the
  /// field would change what the parity test compares.
  public let detail: String?
  public let closureCode: String?
  public let detailParams: [String: String]?
  /// The catalog key for the state's sentence, set on the APP's path (`DayState.ribbonInput`)
  /// and absent on the wire. This is what replaces `detail` for anything a reader sees.
  public let labelKey: String?

  enum CodingKeys: String, CodingKey {
    case facility
    case status
    case detail
    case closureCode = "closure_code"
    case detailParams = "detail_params"
    case labelKey = "label_key"
  }

  public init(
    facility: String?,
    status: String?,
    detail: String?,
    closureCode: String? = nil,
    detailParams: [String: String]? = nil,
    labelKey: String? = nil
  ) {
    self.facility = facility
    self.status = status
    self.detail = detail
    self.closureCode = closureCode
    self.detailParams = detailParams
    self.labelKey = labelKey
  }
}

// MARK: - The mapping

/// One lane's holds, clipped to `[start, end)`.
///
/// The clip is not cosmetic. A `lane_day_view` spans the whole WEEKDAY while a ribbon is one
/// SESSION, and two of a pool's sessions share a basin and therefore share a day view — so an
/// unclipped stack would paint each session across the whole day, two ribbons claiming hours
/// neither covers, drawn over each other.
///
/// A lane with nothing inside the window KEEPS its empty sub-row: "nobody holds this lane" is a
/// fact, and dropping the row would silently renumber the lanes below it.
public func laneStack(
  _ dayView: RibbonDayViewInput,
  from start: String?,
  to end: String?
) -> [RibbonStackLane] {
  guard let lower = TimeOfDay(hhmm: start ?? ""), let upper = TimeOfDay(hhmm: end ?? "") else {
    return []
  }
  var holdsByLane: [Int: [RibbonDayViewInput.Segment]] = [:]
  for strip in dayView.strips ?? [] { holdsByLane[strip.lane] = strip.segments ?? [] }
  guard dayView.laneCount >= 1 else { return [] }
  return (1...dayView.laneCount).map { lane in
    RibbonStackLane(
      lane: lane,
      segments: (holdsByLane[lane] ?? []).compactMap { clipBlock($0, lower, upper) }
    )
  }
}

private func clipBlock(
  _ segment: RibbonDayViewInput.Segment,
  _ lower: TimeOfDay,
  _ upper: TimeOfDay
) -> RibbonStackBlock? {
  guard let holdStart = TimeOfDay(hhmm: segment.start), let holdEnd = TimeOfDay(hhmm: segment.end)
  else { return nil }
  let from = max(lower, holdStart)
  let to = min(upper, holdEnd)
  // Wholly outside this session, or zero-length once clipped.
  guard from < to else { return nil }
  return RibbonStackBlock(
    start: from.hhmm,
    end: to.hhmm,
    isPublic: segment.access == publicSwimKind,
    // An empty owner string is no owner: it would render as an anonymous reservation label.
    owner: (segment.owner?.isEmpty ?? true) ? nil : segment.owner
  )
}

/// Public fraction of a timeline segment. Explicitly 0 — never a division by zero — when the
/// basin records no lanes, so the ribbon pinches shut rather than vanishing into NaN.
func publicFraction(_ segment: RibbonTimelineInput.Segment) -> Double {
  segment.laneCount > 0 ? Double(segment.publicLanes) / Double(segment.laneCount) : 0
}

/// The ribbon for one option — the three-way choice at the heart of this module.
public func optionRibbon(_ option: RibbonOptionInput) -> Ribbon {
  let family = accessFamily(option.access ?? "")
  // Hours are required to CLIP the day view to this session, so an option missing them cannot
  // be stacked: a stack with no window would paint every lane empty — "nothing free" — which is
  // the one thing a missing fact must never look like.
  let hasHours =
    TimeOfDay(hhmm: option.start ?? "") != nil && TimeOfDay(hhmm: option.end ?? "") != nil
  // `strips` must be PRESENT but may be empty, exactly as the browser tests it
  // (`Array.isArray(dayView.strips)`): a plan with a lane count and no strips is still a plan,
  // and its lanes render as empty sub-rows rather than falling back to "not published".
  if let dayView = option.laneDayView, dayView.laneCount > 0, dayView.strips != nil, hasHours {
    return Ribbon(
      kind: "option", variant: "lanestack", style: "solid", family: family,
      access: option.access, facility: option.facility, basin: option.basin,
      start: option.start, end: option.end, sheath: true, laneCount: dayView.laneCount,
      strips: laneStack(dayView, from: option.start, to: option.end), segments: nil,
      // ABSENT, not zero-width, when the server has no window for this session: the band is a
      // claim ("come then"), and a null one is no claim at all.
      bestPublic: option.laneBestPublic, labelKey: nil, detail: nil, closureCode: nil,
      detailParams: nil, status: nil
    )
  }
  if let segments = option.laneTimeline?.segments, !segments.isEmpty {
    return Ribbon(
      kind: "option", variant: "lanes", style: "solid", family: family,
      access: option.access, facility: option.facility, basin: option.basin,
      start: option.start, end: option.end, sheath: true, laneCount: nil, strips: nil,
      segments: segments.map(ribbonSegment), bestPublic: nil, labelKey: nil, detail: nil,
      closureCode: nil, detailParams: nil, status: nil
    )
  }
  return Ribbon(
    kind: "option", variant: "unpublished", style: "solid", family: family,
    access: option.access, facility: option.facility, basin: option.basin,
    start: option.start, end: option.end, sheath: false, laneCount: nil, strips: nil,
    segments: nil, bestPublic: nil, labelKey: noSplitLabelKey, detail: nil, closureCode: nil,
    detailParams: nil, status: nil
  )
}

private func ribbonSegment(_ segment: RibbonTimelineInput.Segment) -> RibbonSegment {
  RibbonSegment(
    start: segment.start,
    end: segment.end,
    thickness: publicFraction(segment),
    pinched: (segment.reservedLanes ?? 0) > 0,
    laneCount: segment.laneCount,
    publicLanes: segment.publicLanes,
    reservedLanes: segment.reservedLanes ?? 0,
    partial: segment.partial
  )
}

/// The ribbon for one status: dashed when the source says closed, dotted otherwise.
///
/// ANY unrecognised status falls to the ghost, never to closed. That direction is the whole
/// point — a client reading a store newer than itself must degrade to "we do not know", which
/// is true, rather than to "it is shut", which would be a fabricated closure.
public func statusRibbon(_ status: RibbonStatusInput) -> Ribbon {
  let closed = status.status == "closed"
  return Ribbon(
    kind: "status",
    variant: closed ? "closed" : "ghost",
    style: closed ? "dashed" : "dotted",
    family: closed ? "closed" : "unknown",
    access: nil, facility: status.facility, basin: nil, start: nil, end: nil, sheath: nil,
    laneCount: nil, strips: nil, segments: nil, bestPublic: nil, labelKey: status.labelKey,
    detail: status.detail, closureCode: status.closureCode, detailParams: status.detailParams,
    // Carried on the ghost so the canvas can render the SPECIFIC freshness state
    // (`awaiting_scrape` vs `no_source`) rather than one merged bucket.
    status: closed ? nil : status.status
  )
}
